packages/google_sign_in/android/src/main/java/io/flutter/plugins/googlesignin/GoogleSignInPlugin.java

```diff
@@ -488,147 +392,104 @@ private void finishWithError(String errorCode, String errorMessage) {
+      if (email == null) {
+        result.error(ERROR_REASON_EXCEPTION, "Email is null", null);
+        return;
       }
 
-      @Override
-      public void onActivityStarted(Activity activity) {
-        if (!resolvingError && activity == registrar.activity() && googleApiClient != null) {
-          googleApiClient.connect();
-        }
-      }
+      Callable<String> getTokenTask =
+          new Callable<String>() {
+            @Override
+            public String call() throws Exception {
+              Account account = new Account(email, "com.google");
+              String scopesStr = "oauth2:" + Joiner.on(' ').join(requestedScopes);
+              return GoogleAuthUtil.getToken(registrar.context(), account, scopesStr);
+            }
+          };
 
-      @Override
-      public void onActivityStopped(Activity activity) {
-        if (activity == registrar.activity() && googleApiClient != null) {
-          googleApiClient.disconnect();
-        }
-      }
+      backgroundTaskRunner.runInBackground(
+          getTokenTask,
+          new BackgroundTaskRunner.Callback<String>() {
+            @Override
+            public void run(Future<String> tokenFuture) {
+              try {
+                String token = tokenFuture.get();
+                HashMap<String, String> tokenResult = new HashMap<>();
+                tokenResult.put("accessToken", token);
+                result.success(tokenResult);
+              } catch (ExecutionException e) {
+                if (e.getCause() instanceof UserRecoverableAuthException) {
+                  if (shouldRecoverAuth) {
+                    // Move this to top of getTokens method.
+                    checkAndSetPendingOperation(METHOD_GET_TOKENS, result, email);
+                    recoverAuthFromException((UserRecoverableAuthException) e.getCause());
+                  } else {
+                    result.error(ERROR_USER_RECOVERABLE_AUTH, e.getLocalizedMessage(), null);
+                  }
+                } else {
+                  result.error(ERROR_REASON_EXCEPTION, e.getCause().getMessage(), null);
+                }
+              } catch (InterruptedException e) {
+                result.error(ERROR_REASON_EXCEPTION, e.getMessage(), null);
+                Thread.currentThread().interrupt();
+              }
+            }
+          });
+    }
 
-      /**
-       * Invoked when the GMS client has successfully connected to the GMS server. This signals that
-       * this listener is properly initialized.
-       */
-      @Override
-      public void onConnected(Bundle connectionHint) {
-        // We can get reconnected if, e.g. the activity is paused and resumed.
-        if (pendingOperation != null && pendingOperation.method.equals(METHOD_INIT)) {
-          finishWithSuccess(null);
-        }
+    @Override
+    public boolean onActivityResult(int requestCode, int resultCode, Intent data) {
+      if (pendingOperation == null) {
```

Comment:
There should always be a pending operation with activity result. Check that first.